app/src/org/commcare/engine/cases/query/CaseIndexPrefetchHandler.java

```diff
@@ -0,0 +1,93 @@
+package org.commcare.engine.cases.query;
+
+import org.commcare.cases.model.Case;
+import org.commcare.cases.query.IndexedValueLookup;
+import org.commcare.cases.query.PredicateProfile;
+import org.commcare.cases.query.QueryCacheEntry;
+import org.commcare.cases.query.QueryContext;
+import org.commcare.cases.query.QueryHandler;
+import org.commcare.cases.util.QueryUtils;
+import org.commcare.models.database.user.models.CaseIndexTable;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.trace.EvaluationTrace;
+import org.javarosa.xpath.expr.XPathExpression;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Vector;
+
+/**
+ * This handler detects contexts in which the query is likely to trip many index lookups and will
+ * strategically perform a cache of relevant indexes
+ *
+ * Created by ctsims on 1/25/2017.
+ */
+
+public class CaseIndexPrefetchHandler implements QueryHandler<IndexedValueLookup> {
```

Comment:
This class also only uses the loadIntoIndexTable function of the CaseIndexTable class. Would be nice if this function and above https://github.com/dimagi/commcare-android/pull/1632/files/a2b748808f18b7474dfc8c4828cf1bff717c7f84#r102203627 could be put into an interface so this could be moved into core